keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,520 @@
+        // ActionSucceed, etc.
+        // +optional
+        ContainerExecutionHookStatuses []ContainerExecutionHookStatus
+
+        // Action Summary status
+        // Default is nil
+        // +optional
+        ActionSucceed *bool
+}
+
+// ContainerExecutionHookStatus represents the current state of a hook for a specific container in a pod
+type ContainerExecutionHookStatus struct {
+        // This field is required
+        PodName string
+
+        // This field is required
+        ContainerName string
+
+        ActionStatus *ExecutionHookActionStatus
+ }
+
+type ExecutionHookActionStatus struct {
+        // If not set, it is nil, indicating Action has not started
+        // If set, it means Action has started at the specified time
+	// +optional
+        ActionTimestamp *int64
+
+        // If not set, it is nil, indicating Action is not complete on the specified container in the pod
+        // +optional
+        ActionSucceed *bool
+
+        // The last error encountered when running the action
+	// +optional
+        Error *HookError
+}
+
+type HookError storage.VolumeError
+```
+
+In the ExecutionHookStatus object, there is a list of ContainerExecutionHookStatus for all selected containers in the pods, each ContainerExecutionHookStatus represents the state of the hook on a specific container.
+
+The ActionSucceed field is a summary status field to indicate the final status of the Action in the hook after the commands are run on all selected containers. If not set, this field is nil, meaning the hook controller still needs to wait for the Action commands to finish running on all containers. If set, either true or false, it means the Action commands have finished running (or timeout) on all containers.
+
+
+Here is the definition for PodAction:
+
+```
+// PodAction describes action commands to run on pods/containers based
+// on specified policies.
+// PodAction does not contain information on pods/containers because those are
+// runtime info.
+// PodAction is namespaced
+type PodAction struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+
+        // This is required.
+        Action *ExecutionHookAction
+}
+
+type ExecutionHookAction struct {
+        // This contains the command to run on a container.
+	// The command should be idempotent.
+        // This is required.
+        Action core_v1.Handler
+
+        // +optional
+        ActionTimeoutSeconds *int64
+}
+```
+
+`ExecutionHook` may be used for different cases, such as
+
+* Application-consistency snapshotting
+* Upgrade
+* Rolling upgrade
+* Debugging
+* Prepare for some lifecycle event like a database migration
+* Reload a config file
+* Restart a container
+
+
+The following gives an example of how to use ExecutionHook for application-consistency snapshotting use case.
+
+### Workflow
+
+* Snapshot API should carry two execution hook information (which will reference two PodAction) for quiescing and unquiescing application. 
```

Comment:
So take the application snapshot case as an example, the snapshot controller creates an `ExecutionHook` object referencing a quiescing `PodAction` to trigger the quiescing, and after the snapshot is completed, creates another `ExecutionHook` object referencing a unquiescing `PodAction` to trigger the unquiescing., right?

Reply:
yes,